lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+      throw new ERR_INVALID_ARG_VALUE('type', type, wrongTypeErr);
+    }
+    if (typeof name !== 'string' || !name) {
+      throw new ERR_INVALID_ARG_TYPE('name', ['string'], name);
+    }
+    if (meta !== undefined && typeof meta !== 'object') {
+      throw new ERR_INVALID_ARG_TYPE('meta', ['object', 'undefined'], meta);
+    }
+
+    this.type = type;
+    this.name = name;
+    this.meta = meta;
+
+    // Before acquiring the channel, check if it already exists.
+    const exists = hasChannel(this.channelName);
+    this.#channel = channel(this.channelName);
+
+    // If the channel is new and there are new channel subscribers,
+    // publish the metric to the new metric channel.
+    if (!exists && newMetricChannel.hasSubscribers) {
+      newMetricChannel.publish(this);
+    }
+
+    ObjectFreeze(this);
+  }
+
+  /**
+   * The channel name of the metric.
+   * @property {string} channelName
+   */
+  get channelName() {
+    return `metrics:${this.type}:${this.name}`;
+  }
+
+  /**
+   * The channel for this metric.
+   * @property {Channel} channel
+   */
+  get channel() {
+    return this.#channel;
+  }
+
+  /**
+   * Whether the metric should report values. If there are no subscribers,
+   * metric preparation and report construction can be skipped.
+   * @property {boolean} shouldReport
+   */
+  get shouldReport() {
+    return this.#channel.hasSubscribers;
+  }
+
+  /**
+   * Report a value to the metric.
+   * @param {number} value The value to report.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  report(value, meta) {
+    // Skip report construction if there are no subscribers.
+    if (!this.shouldReport) return;
+    const report = new MetricReport(this.type, this.name, value,
+                                    mixMeta(this.meta, meta));
+    this.#channel.publish(report);
+  }
+}
+
+/**
+ * Represents a snapshot of a value in time. Will report the value every time
+ * reset() is called, or when applyDelta() is called with a non-zero value.
+ */
+class Gauge {
+  /**
+   * The metric to report to.
+   * @property {Metric} metric
+   */
+
+  /**
+   * The value of the gauge.
+   * @property {number} value
+   */
+
+  /**
+   * @param {Metric} metric The metric to report to.
+   */
+  constructor(metric) {
+    if (!(metric instanceof Metric)) {
+      throw new ERR_INVALID_ARG_TYPE('metric', ['Metric'], metric);
+    }
+    this.metric = metric;
+    this.value = 0;
+  }
+
+  /**
+   * Set the gauge value.
+   * @param {number} value The value to set the gauge to.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  reset(value = 0, meta) {
+    this.value = value;
+    this.metric.report(value, meta);
+  }
+
+  /**
+   * Apply a delta to the gauge.
+   * @param {number} value The delta to apply to the gauge.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  applyDelta(value, meta) {
```

Comment:
should `value` be typed checked or are you counting on / allowing for coercion? If the latter, it would be good to add a comment documenting that